Work with history

DmBorozdin opened the pull request:
### Что сделано

- Создан компонент History.
- Создан приватный маршрут history.
- Добавлена кнопка История для авторизованных пользователей.
- Сделана логика добавления в историю с main, search, card компонентов.
- Сделаны редьюсеры для добавления и удаления истории.
- Изменено имя селектора для получения id авторизованного пользователя.
- Созданы новые селекторы

### Чек-лист

- [x] ПР имеет четкое **название**, короткое и описывающее его суть.
- [x] ПР имеет четкое **описание** того, что в рамках него было сделано.
- [x] Я провел **само-ревью** ПРа. Там нет кода, который я забыл удалить, случайно закоммитил и т.п.
- [x] ПР не содержит кода, который не соответствует цели ПРа.

DmBorozdin committed:
Created History component.
Created private history route.
Added history button for auth users in header.
Made the logic of adding to the history from main, search, card components.
Added reducers for adding and deleting history.
Changed name of selector to get auth user id. Created new selectors.

VictorDanilov commented on src/components/search/search.tsx:
```diff
@@ -23,16 +24,24 @@ const Search = () => {
-        DATE_FORMAT
-      )}&checkOut=${value.rangePicker[1].format(DATE_FORMAT)}`
-    );
+    const url = `${APPRoute.SEARCH}?location=${
+      value.location
+    }&checkIn=${value.rangePicker[0].format(
+      DATE_FORMAT
+    )}&checkOut=${value.rangePicker[1].format(DATE_FORMAT)}`;
+    const locationFormated =
+      value.location[0].toUpperCase() + value.location.slice(1);
+    if (authUser) {
+      dispatch(addHistory({ title: locationFormated, url }));
+    }
+    navigate(url);
```
дублирование с мейн какое-то.

VictorDanilov commented on src/components/main/main.tsx:
```diff
@@ -3,21 +3,33 @@ import styles from "./main.module.scss";
-        DATE_FORMAT
-      )}&checkOut=${value.rangePicker[1].format(DATE_FORMAT)}`
-    );
+    const url = `${APPRoute.SEARCH}?location=${
+      value.location
+    }&checkIn=${value.rangePicker[0].format(
+      DATE_FORMAT
+    )}&checkOut=${value.rangePicker[1].format(DATE_FORMAT)}`;
+    const locationFormated =
+      value.location[0].toUpperCase() + value.location.slice(1);
```
кишки какие-то. Надо прятать.

VictorDanilov commented on src/components/app/App.tsx:
```diff
@@ -31,6 +32,9 @@ const App = () => {
           <Route element={<ProtectedRoute authUser={authUser} />}>
             <Route path={APPRoute.FAVORITES} element={<Favorites />} />
           </Route>
+          <Route element={<ProtectedRoute authUser={authUser} />}>
```
Прям так надо этого authUser прокидывать именно тут, подписывая весь Апп на кусок стейта?

DmBorozdin committed:
Refactoring.
Created SearchForm component.
Moved duplicate code from Search and Main components to SearchForm component.
Moved state subscription from app component to ProtectedRoute component
